contrib/pinot-druid-migration/src/main/java/org/apache/pinot/druid/data/readers/DruidSegmentRecordReader.java

```diff
@@ -0,0 +1,252 @@
+import org.joda.time.chrono.ISOChronology;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The DruidSegmentRecordReader allows us to convert all of the rows in a Druid segment file
+ * into GenericRows, which are made into Pinot segments.
+ *
+ * Note that Druid uses LONG and not INT, so construct the Pinot schema accordingly.
+ */
+public class DruidSegmentRecordReader implements RecordReader {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DruidSegmentRecordReader.class);
+
+  private Schema _pinotSchema;
+  private ArrayList<String> _columnNames;
+  private Cursor _cursor;
+  private List<BaseObjectColumnValueSelector> _selectors;
+  private QueryableIndex _index;
+
+  public DruidSegmentRecordReader(@Nonnull File indexDir, @Nullable Schema schema)
+      throws IOException {
+    init(indexDir, schema);
+  }
+
+  private void init(File indexDir, Schema schema)
+      throws IOException {
+    // Only the columns whose names are in the Pinot schema will get processed
+    _pinotSchema = schema;
+
+    ColumnConfig config = new DruidProcessingConfig() {
+      @Override
+      public String getFormatString() {
+        return "processing-%s";
+      }
+
+      @Override
+      public int intermediateComputeSizeBytes() {
+        return 100 * 1024 * 1024;
+      }
+
+      @Override
+      public int getNumThreads() {
+        return 1;
+      }
+
+      @Override
+      public int columnCacheSizeBytes() {
+        return 25 * 1024 * 1024;
+      }
+    };
+
+    ObjectMapper mapper = new DefaultObjectMapper();
+    final IndexIO indexIO = new IndexIO(mapper, config);
+    _index = indexIO.loadIndex(indexDir);
+    QueryableIndexStorageAdapter adapter = new QueryableIndexStorageAdapter(_index);
+
+    // A Sequence "represents an iterable sequence of elements. Unlike normal Iterators however, it doesn't expose
+    // a way for you to extract values from it, instead you provide it with a worker (an Accumulator) and that defines
+    // what happens with the data."
+    final Sequence<Cursor> cursors = adapter.makeCursors(
+        Filters.toFilter(null),
+        _index.getDataInterval().withChronology(ISOChronology.getInstanceUTC()),
+        VirtualColumns.EMPTY,
+        Granularities.ALL,
+        false,
+        null
+    );
+
+    List<Cursor> cursorList = new ArrayList<>();
+    final Sequence<Object> sequence = Sequences.map(cursors, new Function<Cursor, Object>() {
+      @Override
+      public Object apply(@Nullable Cursor cursor) {
+        cursorList.add(cursor);
+        return null;
+      }
+    });
+    sequence.accumulate(null, (accumulated, in) -> null);
+
+    // There should only be one single Cursor for every segment, so there should only be one Cursor in the cursorList
+    Preconditions.checkArgument(cursorList.size() == 1, "There should only be one Cursor in the Sequence.");
+    _cursor = cursorList.get(0);
+
+    _columnNames = new ArrayList<>();
```

Comment:
Do we need to save this as an instance variable since we can easily access to this by calling `_pinotSchema.getColumnNames()`?

Reply:
Sorry. I actually put the same comment above and you added the answer on why you had to keep `_columnNames`. Resolving this.